### ISSUE TITLE
AWS NLB created without k8s tags on target groups

### ISSUE THREAD

**Issue by masterzen** [kubernetes/kubernetes#73479]
**What happened**:

We created a LoadBalancer service with type 'nlb' with two listeners forwarding traffic on two different ports. The service was created correctly along with the target groups and was functionning correctly.

Around 20 hours later, no traffic was flowing from clients to any k8s nodes. While troubleshooting we discovered that the k8s node instance security group was lacking the input rules allowing traffic to the node ports for this LB.

We deleted the service and recreated it. The service started working normally again, but we found out that the two target groups were missing their k8s tags. 

The logs contain the following error:
~~~
E0129 14:47:01.381032       1 service_controller.go:219] error processing service scalable/pr-testing (will retry): failed to ensure load balancer for service scalable/pr-testing: Error adding tags after modifying load balancer targets: "ValidationError: Only one resource can be tagged at a time\n\tstatus code: 400, request id: bc7a47eb-23d4-11e9-9920-b76ff2fbca55"
I0129 14:47:01.381136       1 event.go:218] Event(v1.ObjectReference{Kind:"Service", Namespace:"scalable", Name:"pr-testing", UID:"b84f1aaf-23d4-11e9-8b3c-0a73252851de", APIVersion:"v1", ResourceVersion:"120452349", FieldPath:""}): type: 'Warning' reason: 'CreatingLoadBalancerFailed' Error creating load balancer (will retry): failed to ensure load balancer for service scalable/pr-testing: Error adding tags after modifying load balancer targets: "ValidationError: Only one resource can be tagged at a time\n\tstatus code: 400, request id: bc7a47eb-23d4-11e9-9920-b76ff2fbca55"
~~~

**What you expected to happen**:

We expected the NLB target groups to have k8s tags, so that k8s could remember it created them.

**How to reproduce it (as minimally and precisely as possible)**:

Create a k8s service with NLB:
~~~yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "true"
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-interval: "300"
    service.beta.kubernetes.io/aws-load-balancer-type: nlb
  labels:
    app: pr-testing
  name: pr-testing
spec:
  externalTrafficPolicy: Cluster
  ports:
  - name: non-ssl
    port: 2345
    protocol: TCP
    targetPort: 2345
  - name: ssl
    port: 2445
    protocol: TCP
    targetPort: 2445
  selector:
    app: pr-testing
  sessionAffinity: None
  type: LoadBalancer
~~~

Add a simple deployment that is matched by this service.

**Anything else we need to know?**:

Looking at AWS cloudtrail we could see the following error event:
~~~json
{
    "eventVersion": "1.05",
    "userIdentity": {
       // elided for security reasons
    },
    "eventTime": "2019-01-29T14:47:01Z",
    "eventSource": "elasticloadbalancing.amazonaws.com",
    "eventName": "AddTags",
    "awsRegion": "us-east-1",
    "sourceIPAddress": "34.201.38.207",
    "userAgent": "aws-sdk-go/1.12.7 (go1.9.3; linux; amd64)",
    "errorCode": "ValidationException",
    "errorMessage": "Only one resource can be tagged at a time",
    "requestParameters": {
        "tags": [
            {
                "value": "kubernetes",
                "key": "KubernetesCluster"
            },
            {
                "value": "owned",
                "key": "kubernetes.io/cluster/kubernetes"
            },
            {
                "value": "scalable/pr-testing",
                "key": "kubernetes.io/service-name"
            }
        ],
        "resourceArns": [
            "arn:aws:elasticloadbalancing:us-east-1:158809441554:targetgroup/k8s-tg-67535809ebce3-2345-30034/8051fd2606c91a15",
            "arn:aws:elasticloadbalancing:us-east-1:158809441554:targetgroup/k8s-tg-67535809ebce3-2445-31527/f32c335558de58e7"
        ]
    },
    "responseElements": null,
    "requestID": "bc7a47eb-23d4-11e9-9920-b76ff2fbca55",
    "eventID": "ec99d2d9-c6be-419c-93e1-eeb7dc81a13a",
    "eventType": "AwsApiCall",
    "apiVersion": "2015-12-01",
    "recipientAccountId": "158809441554"
}
~~~

So the `AddTags` API request was called with two Resource ARNs, which is confirmed by the code:
https://github.com/kubernetes/kubernetes/blob/30ed3b41b7b080cf37639371726a445032d5ef29/pkg/cloudprovider/providers/aws/aws_loadbalancer.go#L309-L312

Since we have two target groups, there are two resources to tag.
Except that the AWS API while allowing to send more than one resource ARNs doesn't allow it returning a validation error.

This could be solved by doing one `AddTags` per target groups.

Note that this can't happen on a Classic LB because there's no target groups.

I believe that those missing tags on the target groups are the root cause of the deletion of the security group rules allowing traffic to reach the nodeports, with this (hypothetical and unproved) scenario:

1. k8s creates the target groups, the NLB, the listeners, the sg input rules
2. k8s fails to tag the target groups, but doesn't rollback anything
3. in a subsequent convergence loops, k8s notices that there are two k8s-managed Security Group input rules, but those don't match any Target Groups or match Target Groups that are not managed by k8s (because they don't have the owner tags)
4. k8s deletes those rules

Unfortunately we don't see the rule deletion in the controller manager logs, but it appears in Cloudtrail:
~~~json
{
    "eventVersion": "1.05",
    "userIdentity": {
       // elided for security reasons
    },
    "eventTime": "2019-01-29T10:18:48Z",
    "eventSource": "ec2.amazonaws.com",
    "eventName": "RevokeSecurityGroupIngress",
    "awsRegion": "us-east-1",
    "sourceIPAddress": "34.201.38.207",
    "userAgent": "aws-sdk-go/1.12.7 (go1.9.3; linux; amd64)",
    "requestParameters": {
        "groupId": "sg-b6c7efcc",
        "ipPermissions": {
            "items": [
                {
                    "ipProtocol": "tcp",
                    "fromPort": 32443,
                    "toPort": 32443,
                    "groups": {},
                    "ipRanges": {
                        "items": [
                            {
                                "cidrIp": "10.16.0.0/16",
                                "description": "kubernetes.io/rule/nlb/health=a9acb96a5230711e98bec125cb496f33"
                            }
                        ]
                    },
                    "ipv6Ranges": {},
                    "prefixListIds": {}
                },
                {
                    "ipProtocol": "tcp",
                    "fromPort": 30346,
                    "toPort": 30346,
                    "groups": {},
                    "ipRanges": {
                        "items": [
                            {
                                "cidrIp": "10.16.0.0/16",
                                "description": "kubernetes.io/rule/nlb/health=a9acb96a5230711e98bec125cb496f33"
                            }
                        ]
                    },
                    "ipv6Ranges": {},
                    "prefixListIds": {}
                }
            ]
        }
    },
    "responseElements": {
        "requestId": "3ed62fec-3e79-4be5-a302-33b139afa49d",
        "_return": true
    },
    "requestID": "3ed62fec-3e79-4be5-a302-33b139afa49d",
    "eventID": "e5c573fa-e68d-4e7f-897f-560ec9ee1008",
    "eventType": "AwsApiCall",
    "recipientAccountId": "158809441554"
}
~~~

Let me know if you have any questions,

**Environment**:
- Kubernetes version: 1.10.7
- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release): CoreOS 1967.3.0
- Kernel (e.g. `uname -a`): Linux 4.14.88-coreos
- Install tools: home grown
- Others:

**Comment by M00nF1sh** [comment]
@masterzen 
Hi, the root cause of your problem should be already solved by https://github.com/kubernetes/kubernetes/pull/68422.
It's cherry-picked into old versions and available as v1.11.7, v1.12.5 and v1.13.3(hasn't released yet)

**Comment by masterzen** [comment]
@M00nF1sh ,

Thanks, you're most certainly right about the root cause of the issue being #68422.
It indeed looks exactly like what we've experienced.

Still there's the problem of target groups not being tagged.
